### PR TITLE
Enable cancellation for anonymous pipes and non-async named pipes on Windows

### LINKED ISSUE #31390: Named pipes hang when written asynchronously w/o async flag
When creating named pipes on .NET or .NET Core, unless I use `PipeOptions.Asynchronous`, async I/O operations done later end up hanging the I/O operations. Even if I pass in a `CancellationToken`, these async I/O methods do not cancel when demanded.

IMO the async flag should be simply an optimization -- not something that if set incorrectly causes a hang.
But even if a hang is acceptable, surely the `CancellationToken` should be honored, no?

Minimal repro: [XUnitTestProject5.zip](https://github.com/dotnet/corefx/files/3805221/XUnitTestProject5.zip)
You'll see the test passes, but if you change `PipeOptions` used in the repro to `None`, the test hangs.

Source code: 

```cs
using System;
using System.IO;
using System.IO.Pipes;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

#nullable enable

public class UnitTest1
{
    private readonly CancellationToken TimeoutToken = new CancellationTokenSource(5000).Token;

    private readonly string PipeName = Guid.NewGuid().ToString();
    private const PipeOptions Options = PipeOptions.Asynchronous; // FAILS when set to None
    private const PipeDirection Direction = PipeDirection.InOut;

    [Fact]
    public async Task Test1()
    {
        var serverTask = RunServer();
        using var clientPipe = new NamedPipeClientStream(".", PipeName, Direction, Options);
        await clientPipe.ConnectAsync(this.TimeoutToken);

        Task readTask = ReadBlockAsync(clientPipe, 3, this.TimeoutToken);
        await clientPipe.WriteAsync(new byte[] { 1, 2, 3 }, this.TimeoutToken);
        await clientPipe.FlushAsync(this.TimeoutToken);
        await Task.WhenAll(readTask, serverTask);
    }

    private async Task RunServer()
    {
        using var serverPipe = new NamedPipeServerStream(PipeName, Direction, maxNumberOfServerInstances: 1, PipeTransmissionMode.Byte, Options);
        await serverPipe.WaitForConnectionAsync(this.TimeoutToken);

        Task readTask = ReadBlockAsync(serverPipe, 3, this.TimeoutToken);
        await serverPipe.WriteAsync(new byte[] { 4, 5, 6 }, this.TimeoutToken);
        await serverPipe.FlushAsync(this.TimeoutToken);
        await readTask;
    }

    private static async Task<ReadOnlyMemory<byte>> ReadBlockAsync(Stream stream, int length, CancellationToken cancellationToken)
    {
        byte[] buffer = new byte[length];
        int bytesRead = 0;
        while (bytesRead < length)
        {
            int bytesJustRead = await stream.ReadAsync(buffer.AsMemory(bytesRead), cancellationToken);
            if (bytesJustRead == 0)
            {
                throw new Exception("Lost connection.");
            }

            bytesRead += bytesJustRead;
        }

        return buffer;
    }
}
```

### LINKED ISSUE #63536: Support async NamedPipeServerStream.WaitForConnection for synchronous pipes
Motivating use case: My codebase is using synchronous pipe IO for IPC. The application spends a lot of CPU in the IPC mechanism, and synchronous IO uses less CPU than async IO. Thread count is not an issue, and synchronous code is simpler than async code. Sync IO is an appropriate solution.

I only need async for the `WaitForConnection` call because the child process might die and never connect. A timeout is not possible. Instead, I'm listening for process exit to terminate the pipe.

A synchronous `WaitForConnection` call is really hard to break. For example, closing the pipe handle (`Dispose`) does not unblock that call... Probably, `CancelSynchronousIo` could do it, but I read in [another issue](https://github.com/dotnet/runtime/issues/31390#issuecomment-549585703) that this is hard to implement.

`WaitForConnectionAsync` works on a synchronous pipe, but it simply delegates the sync call to the thread-pool with no ability to cancel. That call will hang forever. I wonder if this implicit async-over-sync design might be a foot gun for people.

Currently, the only workaround that I know of is to create a dummy client pipe and connect to the server to break the `WaitForConnection` call. This scheme is not the nicest solution in the world, and I wonder if there are circumstances where it would break (security?).

What would it take to support this? Does the Win32 API not support overlapped IO on pipes if the pipe mode is synchronous? I could not find such a statement in the documentation. But if that is the case, `SetNamedPipeHandleMode` can temporarily switch the mode to async. Is this feasible to support? This behavior would then be transparent to callers of `WaitForConnectionAsync`.

An alternative solution is to make `IsAsync` settable. `ReadMode` already is settable, so the infrastructure and precedence for calling `SetNamedPipeHandleMode` are there.

### PR DESCRIPTION
**Pull request by stephentoub**
Although ReadAsync, WriteAsync, and WaitForConnectionAsync on pipes all accept a CancellationToken, that token is only usable on Windows for canceling an in-flight operation when the pipe is using overlapped I/O.  If the pipe was created for non-overlapped I/O, as is the case for anonymous pipes and can be the case for named pipes, the token stops being useful for anything other than an up-front cancellation check.

This change fixes that by using CancelSynchronousIo to cancel the synchronous I/O performed as part of these async operations, which are implemented as async-over-sync (queueing to the thread pool a work item that performs the synchronous I/O).

(The Unix implementation already supports cancellation in these situations.)

Fixes https://github.com/dotnet/runtime/issues/31390
Closes https://github.com/dotnet/runtime/issues/63536

While this is implemented differently, I drew inspiration from [poizan42](https://github.com/poizan42)'s experiments in https://github.com/dotnet/runtime/issues/31390#issuecomment-559392741. Thanks!

And even though we're now actually registering for cancellation, performance is actually better even when cancellation is never requested:
|         Method |         Toolchain | Cancelable | Named |      Mean | Ratio | Allocated | Alloc Ratio |
|--------------- |------------------ |----------- |------ |----------:|------:|----------:|------------:|
| ReadWriteAsync | \main\corerun.exe |      False | False | 12.272 us |  1.00 |     400 B |        1.00 |
| ReadWriteAsync |   \pr\corerun.exe |      False | False |  6.597 us |  0.54 |      84 B |        0.21 |
|                |                   |            |       |           |       |           |             |
| ReadWriteAsync | \main\corerun.exe |      False |  True | 32.442 us |  1.00 |     400 B |        1.00 |
| ReadWriteAsync |   \pr\corerun.exe |      False |  True | 29.737 us |  0.93 |     109 B |        0.27 |
|                |                   |            |       |           |       |           |             |
| ReadWriteAsync | \main\corerun.exe |       True | False | 12.581 us |  1.00 |     400 B |        1.00 |
| ReadWriteAsync |   \pr\corerun.exe |       True | False |  8.036 us |  0.62 |      85 B |        0.21 |
|                |                   |            |       |           |       |           |             |
| ReadWriteAsync | \main\corerun.exe |       True |  True | 31.889 us |  1.00 |     400 B |        1.00 |
| ReadWriteAsync |   \pr\corerun.exe |       True |  True | 29.458 us |  0.92 |     115 B |        0.29 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.IO.Pipes;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private Stream _server;
    private Stream _client;
    private byte[] _buffer = new byte[1];
    private CancellationTokenSource _cts = new CancellationTokenSource();

    [Params(false, true)]
    public bool Cancelable { get; set; }

    [Params(false, true)]
    public bool Named { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        if (Named)
        {
            string name = Guid.NewGuid().ToString("N");
            var server = new NamedPipeServerStream(name, PipeDirection.Out);
            var client = new NamedPipeClientStream(".", name, PipeDirection.In);
            Task.WaitAll(server.WaitForConnectionAsync(), client.ConnectAsync());
            _server = server;
            _client = client;
        }
        else
        {
            var server = new AnonymousPipeServerStream(PipeDirection.Out);
            var client = new AnonymousPipeClientStream(PipeDirection.In, server.ClientSafePipeHandle);
            _server = server;
            _client = client;
        }
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _server.Dispose();
        _client.Dispose();
    }

    [Benchmark(OperationsPerInvoke = 1000)]
    public async Task ReadWriteAsync()
    {
        CancellationToken ct = Cancelable ? _cts.Token : default;
        for (int i = 0; i < 1000; i++)
        {
            ValueTask<int> read = _client.ReadAsync(_buffer, ct);
            await _server.WriteAsync(_buffer, ct);
            await read;
        }
    }
}
```

### CODE REVIEW

**Comment by danmoseley** [comment]
Do we have this covered in the perf benchmarks? If not maybe we could reuse what you have there.

**Review by AArnott — APPROVED** [review]

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs [hunk 1]
```diff
@@ -230,6 +211,187 @@ public override void EndWrite(IAsyncResult asyncResult)
...
+                finally
+                {
+                    // Prior to calling Dispose on the CancellationTokenRegistration, we need to tell
+                    // the registration callback to exit if it's currently running; otherwise, we could deadlock.
+                    workItem.ContinueTryingToCancel = false;
+                }
+            }
+        }
+
+        /// <summary>State object used for <see cref="AsyncOverSyncRead"/> and <see cref="AsyncOverSyncWrite"/>.</summary>
```
Comment (adamsitnik) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs:285]: nit: and `WaitForConnection`

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs [hunk 2]
```diff
@@ -230,6 +211,187 @@ public override void EndWrite(IAsyncResult asyncResult)
...
+            {
+                // If the token can't be canceled, there's nothing to register.
+                if (!cancellationToken.CanBeCanceled)
+                {
+                    return default;
+                }
+
+                // Get a handle for the current thread. This is stored and used to cancel the I/O on this thread
+                // in response to the cancellation token having cancellation requested.  If the handle is invalid,
+                // which could happen if OpenThread fails, skip attempts at cancellation.
```
Comment (adamsitnik) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs:330]: nit: you could mention that `THREAD_TERMINATE` access right is required by the `CancelSynchronousIo` sys-call

**Review by adamsitnik — APPROVED** [review]
It's a really smart way of solving this problem! Thank you for adding all the comments @stephentoub !

We need to figure out a way to reuse the logic for `FileStream`, `RandomAccess` and `Pipes`. And ideally solve https://github.com/dotnet/runtime/issues/28585 while we are there ;)

**Comment by stephentoub** [comment]
> Do we have this covered in the perf benchmarks

Mostly. The only thing it lacks is passing in a cancellation town. 
https://github.com/dotnet/performance/tree/f7ec27a4da2d61b97184f07989e5d9c6e0c21988/src/benchmarks/micro/libraries/System.IO.Pipes

**Comment by GSPP** [comment]
@stephentoub I've got a question about how this works. What happens in this sequence of events:

- We get directly to before the `CancelSynchronousIo` call
- The IO starts and completes
- The main thread moves on and starts another IO at some other program location
- That other IO now gets canceled

I have fought with this kind of IO cancellation before and was then not able to make it work reliably. Curious to learn how you did that.

**Pull request merged**